### PR TITLE
OpenMP 4.5 backend

### PR DESCRIPTION
**Pull request by mcordery**
Modfied the following:
a) Added proper test for omp_target_memcpy's so they would not fail at run time. 
b) modified const reduce() functions in both standard and location versions to work when
called in a target region.
I thested the above code using the pi_offload test code and it all worked fine.

### CODE REVIEW

File: include/RAJA/policy/openmp/target_reduce.hpp [hunk 1]
```diff
@@ -248,10 +251,13 @@ struct TargetReduce {
...
   {
     using NonConst = typename std::remove_const<decltype(this)>::type;
-    return const_cast<NonConst>(this)->reduce(rhsVal);
+    auto ptr = const_cast<NonConst>(this); 
+    Reducer{}(ptr->val.value,rhsVal);
+    return *this;
   }
 
 private:
+  TargetReduce *ptr2this;
```
Comment (willkill07) [inline include/RAJA/policy/openmp/target_reduce.hpp]: This appears to be unused and would save 8B per Reducer

File: include/RAJA/policy/openmp/target_reduce.hpp [hunk 2]
```diff
@@ -90,7 +90,7 @@ struct Offload_Info {
 
 template <size_t Teams, typename T>
 struct Reduce_Data {
-  T value;
+  mutable T value;
```
Comment (willkill07) [inline include/RAJA/policy/openmp/target_reduce.hpp]: Great catch here. Thanks for finding this

File: cmake/SetupCompilers.cmake [hunk 3]
```diff
@@ -59,6 +59,9 @@ if (CMAKE_CXX_COMPILER_ID MATCHES GNU)
 elseif (CMAKE_CXX_COMPILER_ID MATCHES AppleClang)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
 elseif (CMAKE_CXX_COMPILER_ID MATCHES Clang)
+  set(CMAKE_CXX_STANDARD 11) # C++11...
```
Comment (willkill07) [inline cmake/SetupCompilers.cmake]: Does IBM clang play nicely if this is removed? I integrated a PR so this shouldn't be necessary.

**Comment by mcordery** [comment]
Ah, rats. Thought I got everything. I'll get rid of that.
 
----- Original message -----From: Will Killian <notifications@github.com>To: LLNL/RAJA <RAJA@noreply.github.com>Cc: Matthew Cordery <cordery1@llnl.gov>, Author <author@noreply.github.com>Subject: Re: [LLNL/RAJA] Features/killian4/openmp4 enh reducers (#255)Date: Wed, Jun 7, 2017 10:43 AM 
@willkill07 commented on this pull request.
If ptr2this goes away, I say ship it 
In include/RAJA/policy/openmp/target_reduce.hpp:
>    }  private:+  TargetReduce *ptr2this;
This appears to be unused and would save 8B per Reducer
In include/RAJA/policy/openmp/target_reduce.hpp:
> @@ -90,7 +90,7 @@ struct Offload_Info {  template <size_t Teams, typename T> struct Reduce_Data {-  T value;+  mutable T value;
Great catch here. Thanks for finding this

**Comment by mcordery** [comment]
I need to check to see what exactly is needed. Basically, something in this needed to suppress the addition of
-std=gnu++11 whch conflicts with -std=c++11 and has caused problems for me in the past.
 
----- Original message -----From: Will Killian <notifications@github.com>To: LLNL/RAJA <RAJA@noreply.github.com>Cc: Matthew Cordery <cordery1@llnl.gov>, Author <author@noreply.github.com>Subject: Re: [LLNL/RAJA] Features/killian4/openmp4 enh reducers (#255)Date: Wed, Jun 7, 2017 10:43 AM 
@willkill07 commented on this pull request.
If ptr2this goes away, I say ship it 
In include/RAJA/policy/openmp/target_reduce.hpp:
>    }  private:+  TargetReduce *ptr2this;
This appears to be unused and would save 8B per Reducer
In include/RAJA/policy/openmp/target_reduce.hpp:
> @@ -90,7 +90,7 @@ struct Offload_Info {  template <size_t Teams, typename T> struct Reduce_Data {-  T value;+  mutable T value;
Great catch here. Thanks for finding this

**Comment by mcordery** [comment]
It looks like, technically, all that's needed is
  set(CMAKE_CXX_EXTENSIONS OFF)
to keep -std=gnu++14 from being added along with -std=c++11
The compiler works fine now with both but it hasn't in the past and it seems to me that's something we'd want to avoid doing.
 
----- Original message -----From: Matthew Cordery/Almaden/IBMTo: reply@reply.github.comCc:Subject: Re: [LLNL/RAJA] Features/killian4/openmp4 enh reducers (#255)Date: Wed, Jun 7, 2017 10:52 AM 
I need to check to see what exactly is needed. Basically, something in this needed to suppress the addition of
-std=gnu++11 whch conflicts with -std=c++11 and has caused problems for me in the past.
 
----- Original message -----From: Will Killian <notifications@github.com>To: LLNL/RAJA <RAJA@noreply.github.com>Cc: Matthew Cordery <cordery1@llnl.gov>, Author <author@noreply.github.com>Subject: Re: [LLNL/RAJA] Features/killian4/openmp4 enh reducers (#255)Date: Wed, Jun 7, 2017 10:43 AM 
@willkill07 commented on this pull request.
If ptr2this goes away, I say ship it 
In include/RAJA/policy/openmp/target_reduce.hpp:
>    }  private:+  TargetReduce *ptr2this;
This appears to be unused and would save 8B per Reducer
In include/RAJA/policy/openmp/target_reduce.hpp:
> @@ -90,7 +90,7 @@ struct Offload_Info {  template <size_t Teams, typename T> struct Reduce_Data {-  T value;+  mutable T value;
Great catch here. Thanks for finding this

**Comment by mcordery** [comment]
Pushed changes to remove ptr2this and to remove cmake flags that weren't needed for the clang compiler to suppress -std=gnu++14

**Comment by davidbeckingsale** [comment]
I agree with @trws - also, can we try and document the policies before this gets merged? It doesn't need to be anything crazy, just a quick Doxygen-style comment....

**Comment by willkill07** [comment]
Brief, extremely mild documentation is now added. I also moved the reducers to the `pattern/reduce.hpp` header for future adoption/integration with the other reducers (including the `omp_reduce` cruft.

Since these are capable of extending to other policies, e.g. `OmpReduce`/`OmpReduceLoc` or `CudaReduce`/`CudaReduceLoc` I think this is good to go for the time begin.

Let me know your thoughts/additional review critique.

@mcordery thanks for fixing this and pushing it through to be integrated!

**Comment by willkill07** [comment]
Edit: I will fix the conflicts shortly

**Comment by willkill07** [comment]
Changes integrated -- And it still runs on manta!

```
bash-4.2$ ./examples/example-raja-pi-offload
PI is ~ 3.14159
PImin is ~ -0.999996
PImax is ~ 0.999996
PIminloc is ~ -0.999996 at 262143
PImaxloc is ~ 0.999996 at 262143
```

This should be good to go :shipit:

**Comment by mcordery** [comment]
Sweet. I'm working on trying to verify that it works with XL. Ettore put in a fix but it's on another machine and requires linking to something so I'll let you know.
 
----- Original message -----From: Will Killian <notifications@github.com>To: LLNL/RAJA <RAJA@noreply.github.com>Cc: Matthew Cordery <cordery1@llnl.gov>, Mention <mention@noreply.github.com>Subject: Re: [LLNL/RAJA] Features/killian4/openmp4 enh reducers (#255)Date: Fri, Jun 9, 2017 2:21 PM 
Changes integrated -- And it still runs on manta!
bash-4.2$ ./examples/example-raja-pi-offloadPI is ~ 3.14159PImin is ~ -0.999996PImax is ~ 0.999996PIminloc is ~ -0.999996 at 262143PImaxloc is ~ 0.999996 at 262143
This should be good to go

File: include/RAJA/pattern/reduce.hpp [hunk 4]
```diff
@@ -64,6 +65,67 @@
 namespace RAJA
 {
 
+namespace reduce
```
Comment (davidbeckingsale) [inline include/RAJA/pattern/reduce.hpp]: I'm not against this change, but I would like to quantify the performance impact of having these operator structs.

Comment (willkill07) [inline include/RAJA/pattern/reduce.hpp]: Would assembly output be enough to quantify the overhead vs manual/macro alternative?

Comment (willkill07) [inline include/RAJA/pattern/reduce.hpp]: It's also in a common location so other reducers could be restructured to adhere to a similar design. (and leverage OpenMP locks -- something that @trws would certainly prefer as well)

**Comment by trws** [comment]
Yes, very much so.  We have a lot more contention than we need.  The only trick here may be dealing with the host/device split, I'm not sure of the best way to deal with that off top of my head.

and leverage OpenMP locks -- something that @trws<https://github.com/trws> would certainly prefer as well



---
Sent from VMware Boxer

On June 13, 2017 at 9:32:35 AM PDT, Will Killian <notifications@github.com> wrote:

@willkill07 commented on this pull request.

________________________________

In include/RAJA/pattern/reduce.hpp<https://github.com/LLNL/RAJA/pull/255#discussion_r121730154>:

> @@ -64,6 +65,67 @@
 namespace RAJA
 {

+namespace reduce


It's also in a common location so other reducers could be restructured to adhere to a similar design. (and leverage OpenMP locks -- something that @trws<https://github.com/trws> would certainly prefer as well)

**Comment by mcordery** [comment]
Just FYI, I just tested a new IPA library for XLC on one of the IBM clusters and the raja-pi-offload test passed with that.

**Comment by willkill07** [comment]
Performance update:

Reduce Pi Example:

* original took 343us on average (across 10 runs with minimal variance).
* this version took 295us on average (same # of runs w/ similar variance).

Bumping the iterations from 512^2 to 512^3 produced the following times:

* original: 16ms on average
* updated: 4.6ms on average

No, I can't really explain the difference -- cuobjdump doesn't work with binaries emitted from clang++gpu, but I am certainly not complaining.

I was going to use the stress test from @trws but there were only two more reducers being used (all of ReduceSum). The pi offload test does 5 reductions all of different types.

**Comment by willkill07** [comment]
Also, it appears as though this PR addresses the following Issues/PRs:

closes #116
closes #179
closes #247

File: cmake/SetupPackages.cmake [hunk 5]
```diff
@@ -46,6 +46,12 @@ if (RAJA_ENABLE_OPENMP)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
     list(APPEND RAJA_NVCC_FLAGS -Xcompiler ${OpenMP_CXX_FLAGS})
     message(STATUS "OpenMP Enabled")
+    if (RAJA_ENABLE_TARGET_OPENMP)
+      set(RAJA_ENABLE_TARGET_OPENMP On)
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DRAJA_ENABLE_TARGET_OPENMP")
```
Comment (willkill07) [inline cmake/SetupPackages.cmake]: Changes made as requested

Comment (davidbeckingsale) [inline cmake/SetupPackages.cmake]: This stuff doesn't need to be in the CMakeLists.txt anymore then. 😄

Comment (willkill07) [inline cmake/SetupPackages.cmake]: Ugh, you're right >.< fixing

**Comment by willkill07** [comment]
Notes from RAJA meeting: update all tests.

**Comment by willkill07** [comment]
I added tests but the `forallN` ones complain during linking.

Specifically, it complains that it cannot find the destructor for `ForallN_PeelOuter`

Should I just macro-guard this away since it's a known issue?

**Comment by trws** [comment]
It should be solved by adding an explicit but empty destructor.  This is one of the workarounds we hit at the hackathon.

**Comment by trws** [comment]
Indeed it does.  It looks like both the destructor and a constructor for 
wrapped functor types is missing.  I recall all of this being 
constructors and destructors that would normally be defaulted not being 
generated because the types had explicit constructors.  It's probably 
the destructor on PeelOuter and something like the copy constructor on 
BindFirstArg and IndexTypeConverter.

**Comment by willkill07** [comment]
Alright. I'll tinker around with this and post an update

**Comment by willkill07** [comment]
Here's a cleaned up slightly-more-demanged linker output:


### omp_target on outermost loop:

nvlink error   : Undefined reference to `RAJA::ForallN_PeelOuter<NextExec, Body >::ForallN_PeelOuter(Body const&)` in '/var/tmp/killian4/test-reductions-8603ce-1aba7e.cubin'

nvlink error   : Undefined reference to `RAJA::ForallN_PeelOuter<NextExec, Body >::~ForallN_PeelOuter()` in '/var/tmp/killian4/test-reductions-8603ce-1aba7e.cubin'

WHERE

* NextExec = `RAJA::ForallN_Executor<RAJA::ForallN_PolicyPair<RAJA::seq_exec, RAJA::RangeSegment>, RAJA::ForallN_PolicyPair<RAJA::seq_exec, RAJA::RangeSegment> >`
* Body = `RAJA::ForallN_IndexTypeConverter<Lambda&, int, int, int>`

### omp_target on middle loop:

nvlink error   : Undefined reference to `RAJA::ForallN_PeelOuter<NextExec, Body >::ForallN_PeelOuter(Body const&)` in '/var/tmp/killian4/test-reductions-3dd17d-9bdd07.cubin'
nvlink error   : Undefined reference to `RAJA::ForallN_PeelOuter<NextExec, Body >::~ForallN_PeelOuter()` in '/var/tmp/killian4/test-reductions-3dd17d-9bdd07.cubin'

WHERE

* NextExec = `RAJA::ForallN_Executor<RAJA::ForallN_PolicyPair<RAJA::seq_exec, RAJA::RangeSegment> >`
* Body = `RAJA::ForallN_BindFirstArg_HostDevice<RAJA::ForallN_IndexTypeConverter<Lambda&, int, int, int>, int>`

### omp_target on innermost loop:

nvlink error   : Undefined reference to `RAJA::ForallN_PeelOuter<NextExec, Body >::ForallN_PeelOuter(Body const&)` in '/var/tmp/killian4/test-reductions-d99cb7-ef3681.cubin'
nvlink error   : Undefined reference to `RAJA::ForallN_PeelOuter<NextExec, Body >::~ForallN_PeelOuter()` in '/var/tmp/killian4/test-reductions-d99cb7-ef3681.cubin'

WHERE

* NextExec = `RAJA::ForallN_Executor<>`
* Body = `RAJA::ForallN_BindFirstArg_HostDevice<RAJA::ForallN_BindFirstArg_HostDevice<RAJA::ForallN_IndexTypeConverter<Lambda&, int, int, int>, int>, int>`


## Problems

* `ForallN_PeelOuter` has no matching constructor that accepts only a Body -- How did the compiler arrive at wanting to do that?
* Even with an explicitly added destructor for `ForallN_PeelOuter`, the linker still complains about it. I marked it RAJA_HOST_DEVICE as well

**Comment by trws** [comment]
Should all be good now.  The constructor issues were caused by a bug in firstprivate() codegen, then a bunch of runtime errors caused by array not being mapped, and a change in the capture behavior for lambdas that I regard as a bug we should follow up on.  Either way, this version seems to compile and run all tests with the current coral clang.  Because of the use_device_ptr dance, the unit tests will *not* pass with xlc until one or both of those compilers fix bugs requiring the use of that clause.

File: include/RAJA/pattern/forallN.hpp [hunk 6]
```diff
@@ -164,6 +164,8 @@ struct ForallN_IndexTypeConverter {
   RAJA_HOST_DEVICE
   constexpr ForallN_IndexTypeConverter(Self const &o) : body(o.body) {}
 
+  ~ForallN_IndexTypeConverter() {}
```
Comment (willkill07) [inline include/RAJA/pattern/forallN.hpp]: Build is failing for CUDA. I think this needs to be marked RAJA_HOST_DEVICE

**Comment by willkill07** [comment]
This is now good to go

**Comment by trws** [comment]
I'm good with merging this.  @davidbeckingsale, thoughts?  Also, anyone have an idea why travis is spaced out?

**Comment by davidbeckingsale** [comment]
Yeah I'm happy with merging this. I'm not sure if we need to do anything about forallN at this point? Should we just merge as-is and develop more robust forallN support later?

**Comment by willkill07** [comment]
@davidbeckingsale I would hold off on robust forallN support as we'd need to define a few specialized wrappers and figure out an appropriate way to name team policies.

**Comment by davidbeckingsale** [comment]
That's fine with me, thanks @willkill07 !

**Comment by trws** [comment]
The forallN machinery actually works, it just doesn't have a policy for 
collapse yet, and I'd rather deal with that in the updated forallN than 
take more time extending this version.

**Comment by willkill07** [comment]
Can the shiny "merge pull request" button be pressed, or is there another action item for this PR?

**Comment by trws** [comment]
It's just travis now I think.

**Comment by davidbeckingsale** [comment]
BOOM!

**Comment by davidbeckingsale** [comment]
Thanks @willkill07 & @mcordery

**Comment by willkill07** [comment]
Huge shoutout to @mcordery for getting the ball rolling on this

**Comment by rhornung67** [comment]
👍

**Pull request merged**